From c05f89fbefa858781c0c870df6f7c070f273e8a1 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:00 -0400
Subject: [fs] btrfs: replace error code from btrfs_drop_extents

Message-id: <1407179109-13714-294-git-send-email-sandeen@redhat.com>
Patchwork-id: 87215
O-Subject: [RHEL7.1 PATCH 293/422] btrfs: replace error code from btrfs_drop_extents
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: David Sterba <dsterba@suse.cz>

commit 3f9e3df8da3c51649c15db249978a10f7374236a
Author: David Sterba <dsterba@suse.cz>
Date:   Tue Apr 15 18:50:17 2014 +0200

    btrfs: replace error code from btrfs_drop_extents

    There's a case which clone does not handle and used to BUG_ON instead,
    (testcase xfstests/btrfs/035), now returns EINVAL. This error code is
    confusing to the ioctl caller, as it normally signifies errorneous
    arguments.

    Change it to ENOPNOTSUPP which allows a fall back to copy instead of
    clone. This does not affect the common reflink operation.

    Signed-off-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/file.c b/fs/btrfs/file.c
index cd63bfb..567f5e9 100644
--- a/fs/btrfs/file.c
+++ b/fs/btrfs/file.c
@@ -805,7 +805,7 @@ next_slot:
   if (start > key.offset && end < extent_end) {
    BUG_ON(del_nr > 0);
    if (extent_type == BTRFS_FILE_EXTENT_INLINE) {
-    ret = -EINVAL;
+    ret = -EOPNOTSUPP;
     break;
    }
 
@@ -851,7 +851,7 @@ next_slot:
    */
   if (start <= key.offset && end < extent_end) {
    if (extent_type == BTRFS_FILE_EXTENT_INLINE) {
-    ret = -EINVAL;
+    ret = -EOPNOTSUPP;
     break;
    }
 
@@ -877,7 +877,7 @@ next_slot:
   if (start > key.offset && end >= extent_end) {
    BUG_ON(del_nr > 0);
    if (extent_type == BTRFS_FILE_EXTENT_INLINE) {
-    ret = -EINVAL;
+    ret = -EOPNOTSUPP;
     break;
    }
 
diff --git a/fs/btrfs/ioctl.c b/fs/btrfs/ioctl.c
index e79ff6b..2ad7de9 100644
--- a/fs/btrfs/ioctl.c
+++ b/fs/btrfs/ioctl.c
@@ -3066,7 +3066,7 @@ process_slot:
         new_key.offset + datal,
         1);
     if (ret) {
-     if (ret != -EINVAL)
+     if (ret != -EOPNOTSUPP)
       btrfs_abort_transaction(trans,
         root, ret);
      btrfs_end_transaction(trans, root);
@@ -3141,7 +3141,7 @@ process_slot:
         new_key.offset + datal,
         1);
     if (ret) {
-     if (ret != -EINVAL)
+     if (ret != -EOPNOTSUPP)
       btrfs_abort_transaction(trans,
        root, ret);
      btrfs_end_transaction(trans, root);
-- 
1.7.1